lib/ansible/modules/windows/win_scheduled_task.ps1

```
@@ -165,7 +167,23 @@ try {
         Exit-Json $result
     }
 
-    $principal = New-ScheduledTaskPrincipal -UserId "$user" -LogonType ServiceAccount
+    # Handle RunAs/RunLevel options for the task
+
+    if (!$store_password) {
+        # Create a ScheduledTaskPrincipal for the task to run under
+        $principal = New-ScheduledTaskPrincipal -UserId $user -LogonType S4U -RunLevel $runlevel -Id Author
+        $registerRunOptionParams = @{Principal = $principal}
+    }
+    else {
+        # Specify direct credential and run-level values to add to Register-ScheduledTask
+        $registerRunOptionParams = @{
+            User = $user
+            RunLevel = $runlevel
+        }
+        if ($password) {
```

Should this fail is `$password` is not set, if so you can also simplify the if state on line 221 a tiny bit as well?

---

It shouldn't, in this case - this would allow someone to set only `user` and `runlevel` and not explicitly set `store_password` to no, though it will effectively do just that.

221 definitely has some complicated logic to it - if you can think of a better way to refactor that portion and this one (172), let me know.

---

So looking at the code this will occur when `store_password` is true. I would have thought to store the password you would need the password to be defined, is that not the case?

---

That's true; in this case, it's basically additional logic to handle the same thing.  So `!$store_password` is the same as `$store_password -and !$password`.  I could refactor that if statement to make that more clear, if that helps.

---

I see an issue with my logic, you wouldn't have a password for a service account so what you have is correct